### PR TITLE
chore(express): Refactor `requireAuth` and `clerkMiddleware` middlewares

### PR DESCRIPTION
**Pull request by wobsoriano**
## Description

This PR refactors both `clerkMiddleware` and `requireAuth` Express middlewares to simplify DX and reduce confusion when choosing between middlewares:

1. `clerkMiddleware`

Removed option to pass a handler. It does not do anything at all, which can be replaced by a one-liner from userland:

```ts
// Before
app.use(clerkMiddleware((req, res, next) => {
  // req.auth here is undefined so it's not useful at all
}));

// After
app.use(clerkMiddleware());
app.use((req, res, next) => {
  // do stuff
})
```

2. `requireAuth`

Is now using the same logic as `clerkMiddleware` (use `authenticateRequest` and attach `auth` to `request`), except it will redirect to `signInUrl` if unauthenticated instead of calling `next(new Error())`. So you can have `requireAuth` middleware without `clerkMiddleware`:

```ts
// Before - sends error to error middleware
app.use(clerkMiddleware());
app.get('/protected', requireAuth, (req, res) => {
  res.send('This is a protected route');
});

app.use((err, req, res, next) => {
  if (err instanceof UnauthorizedError) {
    res.status(401).send('Unauthorized');
  } else {
    next(err);
  }
});


// After - redirects to signInUrl

// Apply centralized middleware
app.use(requireAuth());

// Apply middleware to a specific route
app.get('/protected', requireAuth({ signInUrl: '/sign-in }), (req, res) => {
  res.send('This is a protected route');
});
```

Other removed exports:

- `UnauthorizedError` (not used anywhere)
- `ForbiddenError` (not used anywhere)

Documentation for this one is still being worked on, we didn't announce the SDK formally, so I think `minor` update is good. The documentation will have migration guide from `@clerk/clerk-sdk-node` to `@clerk/express`. anyway.

Resolves ECO-201

## Checklist

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.
- [x] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerk/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [x] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

### CODE REVIEW

File: packages/express/src/__tests__/clerkMiddleware.test.ts [hunk 1]
```diff
@@ -83,47 +82,6 @@ describe('clerkMiddleware', () => {
...
-    const customMiddleware: RequestHandler = (_request, response, next) => {
-      response.setHeader('x-custom-middleware', 'custom');
-      return next();
-    };
-
-    const response = await runMiddleware([requireAuth, customMiddleware]).expect(500);
-
-    assertNoDebugHeaders(response);
-    expect(response.header).not.toHaveProperty('x-clerk-auth-custom', 'custom-value');
-  });
```
Comment (wobsoriano) [inline packages/express/src/__tests__/clerkMiddleware.test.ts:125]: Removed this tests as `requireAuth` does not rely on `clerkMiddleware` anymore

Comment (wobsoriano) [inline packages/express/src/__tests__/clerkMiddleware.test.ts:125]: Removed this tests as `requireAuth` does not rely on `clerkMiddleware` anymore

Comment (wobsoriano) [inline packages/express/src/__tests__/clerkMiddleware.test.ts:125]: Removed this tests as `requireAuth` does not rely on `clerkMiddleware` anymore

File: packages/express/src/__tests__/clerkMiddleware.test.ts [hunk 2]
```diff
@@ -83,47 +82,6 @@ describe('clerkMiddleware', () => {
...
-    };
-    const options = { publishableKey: 'pk_test_Y2xlcmsuZXhhbXBsZS5jb20k', enableHandshake: true };
-
-    const response = await runMiddleware(clerkMiddleware(handler, options), {
-      Cookie: '__clerk_db_jwt=deadbeef;',
-    }).expect(200, 'Hello world!');
-
-    expect(response.header).toHaveProperty('x-clerk-auth-custom', 'custom-value');
-    assertSignedOutDebugHeaders(response);
-  });
```
Comment (wobsoriano) [inline packages/express/src/__tests__/clerkMiddleware.test.ts:113]: Removed as we removed option to pass a handler in `clerkMiddleware`

File: packages/express/src/authenticateRequest.ts [hunk 3]
```diff
@@ -95,3 +96,38 @@ const absoluteProxyUrl = (relativeOrAbsoluteUrl: string, baseUrl: string): strin
   }
   return new URL(relativeOrAbsoluteUrl, baseUrl).toString();
 };
+
+export const authenticateAndDecorateRequest = (options: ClerkMiddlewareOptions = {}) => {
+  const clerkClient = options.clerkClient || defaultClerkClient;
+  const enableHandshake = options.enableHandshake || false;
+
+  // eslint-disable-next-line @typescript-eslint/no-misused-promises
+  const middleware: RequestHandler = async (request, response, next) => {
```
Comment (jescalan) [inline packages/express/src/authenticateRequest.ts:105]: Have we covered for the scenario where someone uses both middlewares? A little conditional at the top of this helper to just call next if `auth` is already on the request I think would take care of that. May be worth adding a test for this as well.

**Review by jescalan — COMMENTED** [review]
This is beautiful I love it. Such a quick turnaround!

File: packages/express/src/authenticateRequest.ts (hunk 3 above)
Comment (wobsoriano) [reply packages/express/src/authenticateRequest.ts:109]: Nope, but added just now and a test too!

Comment (jescalan) [reply packages/express/src/authenticateRequest.ts:109]: Do we need to error if there are both? Could we just add a conditional check for `req.auth` at the top of the function instead?

Comment (wobsoriano) [reply packages/express/src/authenticateRequest.ts:109]: Ah I get you now, we want to allow usage of both, but check for `req.auth` first then call `next()` if exists so we dont have to run the same logic etc 👍🏼 Perfect for scenario like

```ts
app.use(clerkMiddleware());

app.get('/protected', requireAuth(), (req, res) => {
  res.send('This is a protected route');
});
```

**Comment by LauraBeatris** [comment]
I love those changes, the DX is much better now ❤️

**Review by jescalan — APPROVED** [review]
This looks great to me! Nice work Rob 👏

**Pull request merged**